### PR TITLE
Configure CI including clippy

### PR DESCRIPTION
**Pull request by tcharding**
Do some non-controversial clippy fixes. PR is now just the patches that were a firm ack below.

### CODE REVIEW

**Comment by stevenroose** [comment]
- 2d5d328 add clippy configuration file: ack, interesting Clippy has a msrv field
- 6d94c16 Remove unneeded 'static lifetime: ack
- 968329f Use struct initialisation shorthand: not a fan of these shorthands, it convolutes field names and local variable names
- e9e33ea Make operator precedence explicit: ack
- cc8bbaa Use byte literal instead of cast: ack
- 4ec4d11 Use iter_mut().enumerate() instead of array access: ack I guess, but that one with take() isn't that clear IMO and it's in a test anyway..
- fa2b19d Remove unneeded references: ack
- 366e410 Remove explicit lifetime: ack
- 44f7f0e Use contains combinatior: eh, you can't write a simple `x < Y | x > Z` anymore in Rust?..
- 1a5b4cf Remove unneeded return statement: ack
- f1c4c7d Use is_empty: ack
- 0c8e968 Remove useless question mark operator: these two changes are not entirely identical, I don't see how clippy things they are better.. other than having 5 less characters, but they are less flexible semantically..
- 0fce400 Remove useless use of vec macro: ack
- 9036eee Use unwrap_or_else with panic in test code: I don't think this is better..
- b233d33 Remove unused lifetime: ack
- 077721c Feature gate generate_in call: ack, this is more a bug than a clippy suggestion I guess..
- 148025a Use hyperlink: ack
- 1c70579 Fix build for serde with no default features: this is also not clippy, right? this seems like a bug too.. ack
- dc19280 Add a test script: ack as I guess this only adds the script and doesn't use it yet?
4882055 (origin/pr/27) Add CI by way of GitHub actions: I don't know github actions too well, but it seems like this enforces zero clippy warnings, which doesn't make sense to me.. clippy gives suggestions not problems

**Comment by tcharding** [comment]
Good to see you bro! Will re-spin with just the acks.

**Comment by tcharding** [comment]
> I don't know github actions too well, but it seems like this enforces zero clippy warnings, which doesn't make sense to me.. clippy gives suggestions not problems

Yes it does. How we solve this will effect how I proceed with this PR. If we don't want to enforce clippy in CI then I can move all the clippy fixes to a separate PR and just do an "add github actions" PR. 

I've done #42 to mimize the scrope, the clippy stuff can come later.

File: clippy.toml [hunk 1]
```diff
@@ -0,0 +1 @@
+msrv = "1.41.1"
```
Comment (stevenroose) [inline clippy.toml:1]: I'm interested what this does exactly.

**Comment by stevenroose** [comment]
utACK 38afbed4a75d6cf5ff82c7e743e6da42e810687e

It's rebased on the github ci changes right? So I guess if you unmark it as draft, it will run CI and I can merge it :) Thanks!

**Comment by tcharding** [comment]
Ah I did the original work with clippy configured for 1.41.1 so I can't guarantee that all the changes are sound and correct for Rust 1.29 and I was too lazy to go back over them all, thought I'd just wait till we moved the MSRV since its in the pipeline already.

**Comment by tcharding** [comment]
Do you want this @stevenroose? If not please close - I'm not fussed either way. Thanks

**Comment by stevenroose** [comment]
If you rebase it, I think I'd ack it :) I'd have to re-review.

**Comment by tcharding** [comment]
No worries, I'll come back to it.

**Pull request closed without merging**